### PR TITLE
Add history service API for reading DLQ tasks

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added and implemented `GetDLQTasks` endpoint to the internal history service.

**Why?**
Operators will need this to inspect DLQ tasks. The existing APIs work with the old DLQ implementation, which, among other differences, do not include message IDs.

**How did you test it?**
There is 100% test coverage for all the new code. Tests are split across unit tests and persistence-layer tests.

**Potential risks**
There is some code that wires up a queue factory that returns a SQL implementation that always returns an error (because it's not implemented). However, this method is not called anywhere in production.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/persistence/client/fault_injection.go [hunk 1]
```diff
@@ -248,6 +248,12 @@ func (d *FaultInjectionDataStoreFactory) NewQueue(queueType persistence.QueueTyp
 	return d.Queue, nil
 }
 
+// NewQueueV2 just returns a queue data-access object without any fault injection. We may add this later, but it is not
+// needed for now.
```
Comment (yycptt) [inline common/persistence/client/fault_injection.go:252]: plz convert to a TODO. and we need to keep track of all those TODOs somewhere.

File: service/history/fx.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ package history
 import (
 	"net"
 
+	"go.temporal.io/server/common/persistence"
```
Comment (yycptt) [inline service/history/fx.go:30]: nit: move to server import section. plz fix other files as well.

File: service/history/fx.go [hunk 3]
```diff
@@ -81,6 +82,7 @@ var Module = fx.Options(
 	fx.Provide(HistoryEngineFactoryProvider),
 	fx.Provide(HandlerProvider),
 	fx.Provide(ServiceProvider),
+	fx.Provide(TaskQueueManagerProvider),
```
Comment (yycptt) [inline service/history/fx.go:85]: add TODO here: it should be provided via persistence factory module.

File: common/persistence/data_interfaces.go [hunk 4]
```diff
@@ -1241,15 +1245,25 @@ type (
 		NextPageToken []byte
 	}
 
+	HistoryTask struct {
+		MessageID int64
```
Comment (yycptt) [inline common/persistence/data_interfaces.go:1249]: MessageMetadata maybe?

File: common/persistence/data_interfaces.go [hunk 5]
```diff
@@ -1226,9 +1226,13 @@ type (
...
 
+	// EnqueueTaskRequest does not include a QueueKey because it does not need the QueueKey.Category field, as that can
+	// already be inferred from the Task field.
 	EnqueueTaskRequest struct {
-		QueueKey QueueKey
-		Task     tasks.Task
+		QueueType     QueueV2Type
+		SourceCluster string
+		TargetCluster string
+		Task          tasks.Task
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/data_interfaces.go:1243]: I agree that it's more intuitive, but I don't want to make it possible for the categories to not match

File: service/history/api/getdlqtasks/api.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+// HistoryTaskQueueManger is a pluggable interface for [persistence.HistoryTaskQueueManger] needed for unit tests.
+type HistoryTaskQueueManger interface {
+	ReadRawTasks(context.Context, *persistence.ReadRawTasksRequest) (*persistence.ReadRawTasksResponse, error)
+}
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/api/getdlqtasks/api.go]: Done; I put the interface there and removed this

File: service/history/api/getdlqtasks/api.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		},
+		PageSize:      int(req.PageSize),
+		NextPageToken: req.NextPageToken,
+	})
+	if err != nil {
+		if errors.Is(err, persistence.ErrReadTasksNonPositivePageSize) {
+			return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("Page size must be positive but was %v", req.PageSize))
+		}
+
+		return nil, serviceerror.NewInternal(fmt.Sprintf("GetDLQTasks failed. Error: %v", err))
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/api/getdlqtasks/api.go]: Fixed

File: service/history/fx.go (hunk 3 above)
Comment (MichaelSnowden) [reply service/history/fx.go]: Done (also added internal task)

File: common/persistence/data_interfaces.go (hunk 4 above)
Comment (MichaelSnowden) [reply common/persistence/data_interfaces.go]: Done

**Review by yycptt — APPROVED** [review]

File: service/history/api/getdlqtasks/api.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+			Category:      category,
+			SourceCluster: req.DlqKey.SourceCluster,
+			TargetCluster: req.DlqKey.TargetCluster,
+		},
+		PageSize:      int(req.PageSize),
+		NextPageToken: req.NextPageToken,
+	})
+	if err != nil {
+		if errors.Is(err, persistence.ErrReadTasksNonPositivePageSize) {
+			return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("Page size must be positive but was %v", req.PageSize))
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/api/getdlqtasks/api.go]: Done

File: common/persistence/history_task_queue_manager.go [hunk 9]
```diff
@@ -102,14 +110,18 @@ func (m *HistoryTaskQueueManager) EnqueueTask(ctx context.Context, request *Enqu
 	}, nil
 }
 
+func (m *HistoryTaskQueueManagerImpl) GetTaskShardID(t tasks.Task) int {
+	return tasks.GetShardIDForTask(t, m.numHistoryShards)
+}
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/history_task_queue_manager.go]: Removed entirely

**Pull request merged**